ArduCopter/precision_landing.cpp

```diff
@@ -22,4 +27,197 @@ void Copter::update_precland()
+        // Therefore just descend for now, and check again later if retry is allowed
+    case AC_PrecLand::PldState::TARGET_FOUND:
+    default:
+        // no action required, target is in sight
+        copter.flightmode->run_land_controller();
+        reset_failed_landing_statemachines();
+        break;
+    }
+}
+
+
+// Target is lost (i.e we had it in sight some time back), this method helps decide on what needs to be done next
+// The chosen action depends on user set landing strictness
+void Mode::PrecLand_StateMachine::target_lost_actions()
+{
+    switch (landing_target_lost_action)
+    {
+    case Landing_Target_Lost_Action::INIT:
+    {
+        // figure out how strict the user is with the landing
+        AC_PrecLand::PldRetryStrictness strictness = copter.precland.get_plnd_retry_strictness();
+        if ((strictness == AC_PrecLand::PldRetryStrictness::NORMAL)
+            || (strictness == AC_PrecLand::PldRetryStrictness::VERY_STRICT)) {
+            // We eventually want to retry landing, but lets descend for sometime and hope the target gets in sight
+            // If not, we will retry landing
+            landing_target_lost_action = Landing_Target_Lost_Action::DESCEND;
+        } else if (strictness == AC_PrecLand::PldRetryStrictness::NOT_STRICT) {
+            // User just wants to land, prec land isn't a priority
+            landing_target_lost_action = Landing_Target_Lost_Action::LAND_VERTICALLY;
+        }
+    }
+        break;
+
+    case Landing_Target_Lost_Action::DESCEND:
+        copter.flightmode->run_land_controller();
+        if (AP_HAL::millis() - copter.precland.get_last_precland_output_ms() >= copter.precland.get_min_retry_time_sec()) {
+            // we have descended for sometime and the target still isn't in sight
+            // lets retry
+            landing_target_lost_action = Landing_Target_Lost_Action::RETRY_LANDING;
+            retry_status = RetryLanding::RETRY_INIT;
+        }
+        break;
+
+    case Landing_Target_Lost_Action::RETRY_LANDING:
+        // retry the landing by going to the last known horizontal position of the target
+        retry_landing();
+        break;
+
+    case Landing_Target_Lost_Action::LAND_VERTICALLY:
+    default:
+        // Just land vertically
+        copter.flightmode->run_land_controller();
+        break;
+    }
+}
+
+// Retry landing based on a previously known location of the landing target
+void Mode::PrecLand_StateMachine::retry_landing()
+{
+    if (_retry_count > copter.precland.get_max_retry_allowed() || copter.precland.get_max_retry_allowed() == 0) {
+        // we have exhausted the amount of times vehicle was allowed to retry landing
+        // do failsafe measure so the vehicle isn't stuck in a constant loop
+        do_failsafe_actions();
+        return;
+    }
+
+    // get the last known location of the landing target
+    Vector3f last_known_loc;
+    copter.precland.get_last_detected_landing_location(last_known_loc);
```

Comment:
This pre-supposes some state.  It's reasonable based on the method name and the description, bit I think it would be a reasonable precaution to make `get_last_detected_landing_location()` boolean and check it here.  Implications of something going wrong are pretty dire.

Potentially the caller could pass the location in if they have it handy and verified (but I don't think that's the case)